picasso/src/main/java/com/squareup/picasso3/MediaStoreRequestHandler.kt

```diff
@@ -0,0 +1,136 @@
+      val exifOrientation = getExifOrientation(requestUri)
+      val mimeType = contentResolver.getType(requestUri)
+      val isVideo = mimeType != null && mimeType.startsWith("video/")
+      if (request.hasSize()) {
+        val picassoKind = getPicassoKind(request.targetWidth, request.targetHeight)
+        if (!isVideo && picassoKind == PicassoKind.FULL) {
+          val source = getSource(requestUri)
+          val bitmap = decodeStream(source, request)
+          signaledCallback = true
+          callback.onSuccess(Result.Bitmap(bitmap, LoadedFrom.DISK, exifOrientation))
+          return
+        }
+        val id = ContentUris.parseId(requestUri)
+        val options = checkNotNull(createBitmapOptions(request), { "options == null" })
+        options.inJustDecodeBounds = true
+        calculateInSampleSize(
+          request.targetWidth, request.targetHeight, picassoKind.width,
+          picassoKind.height, options, request
+        )
+        val bitmap: Bitmap? = if (isVideo) {
+          // Since MediaStore doesn't provide the full screen kind thumbnail, we use the mini kind
+          // instead which is the largest thumbnail size can be fetched from MediaStore.
+          val kind =
+            if (picassoKind == PicassoKind.FULL) Video.Thumbnails.MINI_KIND else picassoKind.androidKind
+          Video.Thumbnails.getThumbnail(contentResolver, id, kind, options)
+        } else {
+          MediaStore.Images.Thumbnails.getThumbnail(
+            contentResolver,
+            id,
+            picassoKind.androidKind,
+            options
+          )
+        }
+        if (bitmap != null) {
+          signaledCallback = true
+          callback.onSuccess(Result.Bitmap(bitmap, LoadedFrom.DISK, exifOrientation))
+          return
+        }
+      }
+      val source = getSource(requestUri)
+      val bitmap = decodeStream(source, request)
+      signaledCallback = true
+      callback.onSuccess(Result.Bitmap(bitmap, LoadedFrom.DISK, exifOrientation))
+    } catch (e: Exception) {
+      if (!signaledCallback) {
+        callback.onError(e)
+      }
+    }
+  }
+
+  override fun getExifOrientation(uri: Uri): Int {
+    var cursor: Cursor? = null
+    return try {
+      val contentResolver = context.contentResolver
+      cursor = contentResolver.query(uri, CONTENT_ORIENTATION, null, null, null)
+      if (cursor == null || !cursor.moveToFirst()) {
+        0
+      } else cursor.getInt(0)
+    } catch (ignored: RuntimeException) {
+      // If the orientation column doesn't exist, assume no rotation.
+      0
+    } finally {
+      cursor?.close()
+    }
+  }
+
+  internal enum class PicassoKind(val androidKind: Int, val width: Int, val height: Int) {
+    MICRO(
+      MediaStore.Images.Thumbnails.MICRO_KIND,
+      96,
+      96
+    ),
+    MINI(MediaStore.Images.Thumbnails.MINI_KIND, 512, 384), FULL(
+      MediaStore.Images.Thumbnails.FULL_SCREEN_KIND, -1, -1
+    );
```

Comment:
```suggestion
    MICRO(MediaStore.Images.Thumbnails.MICRO_KIND, 96, 96),
    MINI(MediaStore.Images.Thumbnails.MINI_KIND, 512, 384), 
    FULL(MediaStore.Images.Thumbnails.FULL_SCREEN_KIND, -1, -1);
```

geez, the auto-formatter really mangled this one...